src/renderer/rendered_window.rs

```diff
@@ -13,21 +13,63 @@ use crate::{
     utils::RingBuffer,
 };
 
-#[derive(Clone, Debug, PartialEq)]
-pub struct LineFragment {
+#[derive(Debug)]
+pub struct Line {
     pub text: String,
-    pub window_left: u64,
-    pub width: u64,
+    pub fragments: Vec<LineFragment>,
+}
+
+#[derive(Debug)]
+pub struct LineFragment {
+    pub text: Range<u32>,
+    pub cells: Range<u32>,
     pub style: Option<Arc<Style>>,
+    pub words: Vec<Word>,
+}
+
+#[derive(Debug, Default)]
+pub struct Word {
+    pub text: u32,
+    pub cell: u32,
+    pub cluster_sizes: Vec<u8>,
+}
+
+pub struct RenderedWord<'a> {
```

Comment:
Its a little weird for RenderedWord to not actually have anything Rendered in it. Would it make sense for this to be ClusteredWord instead?

Reply:
I will try to think of a better name. `Rendered` here refers to that this is the struct actually used when rendering and the `GridRenderer` constructs them. And it will end up being displayed on the screen after shaping and processed by Skia. But yes, it's true that it's not yet renderered.

Something like `ShaperWord` could work since it's the representation used by the shaper. But I don't like that too much either.

Reply:
Maybe RenderableWord?

Reply:
`RenderedWord` is now `Word` and `Word` was renamed to `WordData`. 

I also did some other refactoring related to this by exposing iterator functions instead of having to pass the line string around. See https://github.com/neovide/neovide/pull/3242/commits/2fa2bf3ba413b658a5ef435cdbe8d8553130de5d for the full refactoring